### PR TITLE
LibGUI+LibGfx: Teach fonts Italics

### LINKED ISSUE #9384: LibGfx: Fonts with multiple variants confuse the font picker
*Liberation Serif* is part of the system, as 4 files:

- LiberationSerif-Bold.ttf
- LiberationSerif-BoldItalic.ttf
- LiberationSerif-Italic.ttf
- LiberationSerif-Regular.ttf

But it appears in the font picker dialog like this:
image

Two things are wrong here, probably both part of the same issue:

1. Font size numbers are duplicated for the "Regular" weight. It shows 3 for it, and only 1 for the "Bold" weight, which implies that 3 of the font files are being combined here.
2. The text preview is showing an italic font.

I'd guess the font database is not aware of italic maybe. In any case, it seems to be merging font files together when loading them.

Note that duplicating the Liberation Serif files does not make the problem worse. So only fonts with different attributes get combined like this.

### PR DESCRIPTION
**Pull request by thankyouverycool**
serifina

Fixes #9384
fontpicker

### CODE REVIEW

File: Userland/Libraries/LibGfx/FontStyleMapping.h [hunk 1]
```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright (c) 2020, Andreas Kling <kling@serenityos.org>
```
Comment (alimpfard) [inline Userland/Libraries/LibGfx/FontStyleMapping.h:2]: Did Andreas write this header?

File: Userland/Applications/FontEditor/FontEditor.cpp [hunk 2]
```diff
@@ -495,19 +509,22 @@ void FontEditorWidget::initialize(const String& path, RefPtr<Gfx::BitmapFont>&&
...
-        if (it.weight == m_edited_font->weight()) {
+    for (auto it : Gfx::font_weight_names) {
+        if (it.style == m_edited_font->weight()) {
             m_weight_combobox->set_selected_index(i);
             break;
         }
         i++;
     }
+    i = 0;
+    for (auto it : Gfx::font_slope_names) {
```
Comment (alimpfard) [inline Userland/Applications/FontEditor/FontEditor.cpp:521]: Let's avoid copying this here.

File: Userland/Applications/FontEditor/FontEditor.cpp [hunk 3]
```diff
@@ -415,9 +418,20 @@ FontEditorWidget::FontEditorWidget(const String& path, RefPtr<Gfx::BitmapFont>&&
     };
 
     m_weight_combobox->on_change = [this](auto&, auto&) {
-        m_edited_font->set_weight(GUI::name_to_weight(m_weight_combobox->text()));
+        m_edited_font->set_weight(Gfx::name_to_weight(m_weight_combobox->text()));
```
Comment (alimpfard) [inline Userland/Applications/FontEditor/FontEditor.cpp:421]: What happened to the ::GUI version of this function and its friends?

File: Userland/Applications/FontEditor/FontEditor.cpp [hunk 4]
```diff
@@ -495,19 +509,22 @@ void FontEditorWidget::initialize(const String& path, RefPtr<Gfx::BitmapFont>&&
...
 
-    m_font_weight_list.clear();
-    for (auto& it : GUI::font_weight_names)
-        m_font_weight_list.append(it.name);
-    m_weight_combobox->set_model(*GUI::ItemListModel<String>::create(m_font_weight_list));
-
     int i = 0;
-    for (auto it : GUI::font_weight_names) {
-        if (it.weight == m_edited_font->weight()) {
+    for (auto it : Gfx::font_weight_names) {
```
Comment (alimpfard) [inline Userland/Applications/FontEditor/FontEditor.cpp:513]: Same copying thing here.

File: Userland/Libraries/LibGfx/TrueTypeFont/Font.cpp [hunk 5]
```diff
@@ -462,29 +467,11 @@ String Font::variant() const
...
-    if (variant_name == "Light")
-        return 300;
-    if (variant_name == "Regular")
-        return 400;
-    if (variant_name == "Medium")
-        return 500;
-    if (variant_name == "Semi Bold")
-        return 600;
-    if (variant_name == "Bold")
+    constexpr u16 bold_bit { 0b0000'0000'0000'0001 };
```
Comment (alimpfard) [inline Userland/Libraries/LibGfx/TrueTypeFont/Font.cpp:470]: So...`{ 1 }`?

File: Userland/Libraries/LibGfx/BitmapFont.cpp [hunk 6]
```diff
@@ -348,6 +349,20 @@ String BitmapFont::qualified_name() const
     return String::formatted("{} {} {}", family(), presentation_size(), weight());
 }
 
+String BitmapFont::variant() const
+{
+    StringBuilder variant;
+    variant.append(weight_to_name(weight()));
+    if (slope()) {
+        if (variant.to_string() == "Regular")
```
Comment (alimpfard) [inline Userland/Libraries/LibGfx/BitmapFont.cpp:357]: No need to create a new string, just
```suggestion
        if (variant.string_view() == "Regular"sv)
```

File: Userland/Libraries/LibGfx/BitmapFont.cpp [hunk 7]
```diff
@@ -348,6 +349,20 @@ String BitmapFont::qualified_name() const
...
+            variant.clear();
+        else
+            variant.append(" ");
+        variant.appendff("{}", slope_to_name(slope()));
```
Comment (alimpfard) [inline Userland/Libraries/LibGfx/BitmapFont.cpp:361]: nothing's happening in the format, so 
```suggestion
        variant.append(slope_to_name(slope()));
```

File: Userland/Libraries/LibGfx/BitmapFont.cpp [hunk 8]
```diff
@@ -348,6 +349,20 @@ String BitmapFont::qualified_name() const
     return String::formatted("{} {} {}", family(), presentation_size(), weight());
 }
 
+String BitmapFont::variant() const
+{
+    StringBuilder variant;
```
Comment (alimpfard) [inline Userland/Libraries/LibGfx/BitmapFont.cpp:354]: I'd suggest renaming this to `variant_builder` or just `builder` tbh

File: Userland/Libraries/LibGfx/FontStyleMapping.h (hunk 1 above)
Comment (thankyouverycool) [reply Userland/Libraries/LibGfx/FontStyleMapping.h:2]: Yep, his code was merged into it.

File: Userland/Applications/FontEditor/FontEditor.cpp (hunk 3 above)
Comment (thankyouverycool) [reply Userland/Applications/FontEditor/FontEditor.cpp:421]: These functions come from the now deleted FontPickerWeightModel.h. FontPicker no longer needs that model, but its associated mapping is still useful and used by BitmapFonts, so it was moved to LibGfx in FontStyleMapping.h. Made this more clear in the commit.

**Comment by awesomekling** [comment]
That looks so weird but I can't help but like it.
Maybe we could do programmatic italicization of existing fonts, too? :thonk:

**Comment by thankyouverycool** [comment]
> programmatic italicization

this would be cool! I think that's what "oblique" usually signifies--an algorithmic slant. But genuine, painstakingly bespoke _i t a l i c s_ do be kinda nice

**Pull request merged**